src/Web/Handler.hs

```diff
@@ -147,6 +200,161 @@ getTheory idx = do
     liftIO $ withMVar (theoryVar yesod) $ return. M.lookup idx
 
 
+getLemmaPlaintext :: Int -> TheoryPath -> Handler String
+getLemmaPlaintext nr path = do
+    let lname = case path of 
+            (TheoryEdit n) -> n 
+            _ -> "this should never be visible"
+    eitherTheory <- getTheory nr
+    let lemmaItem = case eitherTheory of 
+            (Just (Trace thy)) -> lookupLemma lname (tiTheory thy)
+            _ -> Nothing
+    let plaintext =  fromMaybe "Enter your new Lemma" $ get lPlaintext <$> lemmaItem
+    return plaintext
+
+
+getStartingProof ::  System -> IncrementalProof
+getStartingProof gsys = LNode (ProofStep (Sorry Nothing) (Just gsys)) M.empty
+
+
+editProof :: Int -> String -> Handler (Either String TheoryIdx)
+editProof idx name = do
+    let result = withTheory idx $ \ti -> do
+            let maybeLemma = lookupLemma name (tiTheory ti)
+            case maybeLemma of
+                Nothing -> return $ Left "Lemma not found"
+                Just (Lemma n m pt tq f a olp) -> do
+                        let ctxt = getProofContext (Lemma n m pt tq f a lp) (tiTheory ti)
+                            preI = getLemmaPreItems n (tiTheory ti)
+                            gsys = mkSystem ctxt (theoryRestrictions (tiTheory ti)) preI f
+                            lp = case olp of
+                                        (LNode (ProofStep Invalidated _) s ) ->
+                                            case M.lookup "" s of
+                                                Just old_lp -> fromMaybe old_lp $ 
+                                                                    runProver (checkAndExtendProver (sorryProver Nothing)) ctxt 0 gsys old_lp
+                                                Nothing -> fromMaybe olp $ runProver (checkAndExtendProver (sorryProver Nothing)) ctxt 0 gsys olp
+
+                                        _ -> fromMaybe olp $ runProver (checkAndExtendProver (sorryProver Nothing)) ctxt 0 gsys olp
+                            editf (Lemma n' pt m tq f a olp) = if n'== n then Lemma n' pt m tq f a lp else Lemma n' pt m tq f a olp
+                            maybe_nthy = modifyLemma editf (tiTheory ti)
+
+                        case maybe_nthy of
+                            Nothing -> return $ Left "Lemma editing failed"
+                            Just nthy -> do
+                                            nidx <- replaceTheory (Just ti) Nothing nthy ("modified" ++ show idx) idx
+                                            return $ Right nidx
```

Comment:
This function uses a lot of nested `Maybe` pattern-matching, which I find hard to follow. I would suggest replacing this with using the Maybe-Monad and using the function `maybe` often. If you want, I can also have a stab at it and submit a PR to your repo.